server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -322,51 +322,55 @@ private static Map<String, Map<String, FieldCapabilities>> buildResponseMap(
         Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder,
         boolean includeUnmapped
     ) {
-        Map<String, Map<String, FieldCapabilities>> responseMap = new HashMap<>();
+        Map<String, Map<String, FieldCapabilities>> responseMap = Maps.newMapWithExpectedSize(responseMapBuilder.size());
+        final var indices = indexResponsesMap.keySet();
         for (Map.Entry<String, Map<String, FieldCapabilities.Builder>> entry : responseMapBuilder.entrySet()) {
-            Map<String, FieldCapabilities.Builder> typeMapBuilder = entry.getValue();
+            var typeMapBuilder = entry.getValue().entrySet();
 
-            Optional<Function<Boolean, FieldCapabilities>> unmapped = Optional.empty();
+            Function<Boolean, FieldCapabilities> unmapped = null;
             if (includeUnmapped) {
                 // do this directly, rather than using the builder, to save creating a whole lot of objects we don't need
                 unmapped = getUnmappedFields(
-                    indexResponsesMap.keySet(),
+                    indices,
                     entry.getKey(),
-                    typeMapBuilder.values().stream().flatMap(FieldCapabilities.Builder::getIndices).collect(Collectors.toSet())
+                    typeMapBuilder.stream().flatMap(t -> t.getValue().getIndices()).collect(Collectors.toSet())
```

Comment:
This is better now but still nasty, doing this kind of stream allocation in a hot loop isn't great.